src/librustc_typeck/diagnostics.rs

```diff
@@ -152,6 +327,20 @@ fn(isize, *const *const u8) -> isize
 ```
 "##,
 
+E0166: r##"
+This error means that the compiler found a return expression in a function
+marked as diverging. A function diverges if it has `!` in the place of the
+return type in its signature. For example:
+
+```
+fn foo() -> ! { return; } // error
+```
+
+For a function that diverges, every control path in the function must end
+with a call to `panic!()` or another diverging function. Attempting to return
```

Comment:
Perhaps this could be re-worded to something like:

> For a function that diverges, every control path in the function must never return, for example with a `loop` that never breaks or a call to another diverging function.

Specifically a call to `panic!()` is just a call to another diverging function, and it'd be nice to mention a concrete case of something that doesn't return as well.